### DISCUSSION TITLE
How to troubleshoot messages not picked up by consumer

### DISCUSSION THREAD

**Discussion by yonigibbs** [nats-io/nats-server#4928] [Q&A]
Hi All,

I am doing some chaos engineering and can pretty reliably replicate a situation where a message in a work queue seems to be stuck, i.e. it's in the queue but not picked up by the queue consumer. Messages before it were all picked up correctly, and newly published messages after it are also being picked up correctly.

To get into this situation I load, say, 100 messages into the queue, then start a JVM process which is processing the items (using Kotlin, in a codebase heavily using coroutines, in case it ends up being relevant). This process picks up multiple messages and processes them concurrently. In that process I cause a `java.lang.OutOfMemoryError` to occur in some thread. In response to this we shut down all running coroutines, NAKing the messages that are in progress, and unsubscribe the JetStream subscription. The JVM process runs for a few seconds, processes some of the items, then dies as expected. If I run it a few times eventually it manages to process all the items in the queue, but one or two are always left in the queue. If I then remove the `OutOfMemoryError` and leave the process up and running, it never picks up the message, no matter how long it's left running.

Because this is simulating a crash of the system, it's possible that there are some race conditions between threads NAKing the messages they're processing, and the subscription being unsubscribed. But I'd expect the system to eventually recover and for the message to eventually be made available to some new consumer that starts up. However this doesn't seem to happen. It seems to remain stuck in the queue.

Does anyone have any ideas how to start troubleshooting this?

Note that if I stop the Java process, remove the consumer, then start up the Java process (which programmatically re-creates the consumer), then this time it does pick up the "stuck" message.

I'm using NATS 2.10-alpine, just running in Docker Compose locally at the moment. Below are some commands I ran that show information that might be pertinent...

## nats stream info
```
$ nats stream info MY-QUEUE
Information for Stream MY-QUEUE created 2024-01-08 16:33:07

           Description: foo bar
              Replicas: 1
               Storage: File

Options:

             Retention: WorkQueue
       Acknowledgments: true
        Discard Policy: Old
      Duplicate Window: 0s
            Direct Get: true
     Mirror Direct Get: true
     Allows Msg Delete: true
          Allows Purge: true
        Allows Rollups: false

Limits:

      Maximum Messages: 100,000,000
   Maximum Per Subject: unlimited
         Maximum Bytes: unlimited
           Maximum Age: unlimited
  Maximum Message Size: unlimited
     Maximum Consumers: unlimited

Replication:

                Mirror: MY-QUEUE-ARCHIVE

Mirror Information:

           Stream Name: MY-QUEUE-ARCHIVE
                   Lag: 0
             Last Seen: 757ms

State:

              Messages: 1
                 Bytes: 189 B
        First Sequence: 71 @ 2024-01-08 16:41:45 UTC
         Last Sequence: 127 @ 2024-01-08 16:41:45 UTC
      Deleted Messages: 56
      Active Consumers: 1
    Number of Subjects: 1
```

## nats consumer info
```
$ nats consumer info MY-QUEUE MY-QUEUE-C-UNKNOWN
Information for Consumer MY-QUEUE > MY-QUEUE-C-UNKNOWN created 2024-01-08T16:33:07Z

Configuration:

                    Name: MY-QUEUE-C-UNKNOWN
             Description: foo bar
               Pull Mode: true
          Filter Subject: my.queue
          Deliver Policy: All
              Ack Policy: Explicit
                Ack Wait: 2m0s
           Replay Policy: Instant
         Max Ack Pending: 1,000
       Max Waiting Pulls: 512
      Inactive Threshold: 5d0h0m0s

State:

  Last Delivered Message: Consumer sequence: 147 Stream sequence: 127 Last delivery: 5m32s ago
    Acknowledgment Floor: Consumer sequence: 147 Stream sequence: 127 Last Ack: 5m31s ago
        Outstanding Acks: 0 out of maximum 1,000
    Redelivered Messages: 0
    Unprocessed Messages: 0
           Waiting Pulls: 1 of maximum 512
```

## nats consumer report
```
$ nats consumer report MY-QUEUE
╭─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                   Consumer report for MY-QUEUE with 1 consumers                                   │
├──────────────────────────────┬──────┬────────────┬──────────┬─────────────┬─────────────┬─────────────┬───────────┬─────────┤
│ Consumer                     │ Mode │ Ack Policy │ Ack Wait │ Ack Pending │ Redelivered │ Unprocessed │ Ack Floor │ Cluster │
├──────────────────────────────┼──────┼────────────┼──────────┼─────────────┼─────────────┼─────────────┼───────────┼─────────┤
│ MY-QUEUE-C-UNKNOWN           │ Pull │ Explicit   │ 2m0s     │ 0           │ 0           │ 0           │ 127       │         │
╰──────────────────────────────┴──────┴────────────┴──────────┴─────────────┴─────────────┴─────────────┴───────────┴─────────╯
```

## nats stream view
```
$ nats stream view MY-QUEUE
[71] Subject: my.queue Received: 2024-01-08T16:41:45Z

{"my":"json"}


16:52:10 Reached apparent end of data
```

(For the latter I had to stop the service and delete the consumer otherwise I hit the error described in this issue: https://github.com/nats-io/natscli/issues/920)

Thanks!

**Comment by yonigibbs** [comment]
Some extra info:
* I notice the `Ack Floor` on the consumer is `127`, but the stream contains a message with sequence number `71`. Should that be possible in a work queue? If the floor is 127 shouldn't message 71 be gone, like all the rest are?
* From our logs it looks like our consumer never received 71. Maybe it received it but the app crashed before it managed to deliver it to our code which logs the sequence number.
* I tried to use the NATS CLI to get at that message. See below...

```
$ nats consumer sub --no-ack MY-QUEUE MY-QUEUE-C-UNKNOWN
nats: error: no message received: nats: timeout
```

**Reply by derekcollison** [reply]
Do you have a max retry set? If that limit is hit (we send an advisory on that condition) the ack floor will move past that message.

**Reply by yonigibbs** [reply]
Pardon my ignorance, but what "max retry" setting specifically do you mean? Am using the Java API if that makes a difference. I can see settings such as `maxDeliver`, `maxAckPending` and `maxPullWaiting` on the consumer. Is it one of those? Or something on the stream itself?

But also, from what I can see in the logs, the message is never even delivered once to the consumer.

**Reply by gregorycollins** [reply]
Max retry is unlimited here

**Reply by gregorycollins** [reply]
P.S. we can trigger this reliably and are working on getting you a minimal example for reproduction.

**Reply by scottf** [reply]
@gregorycollins Is there any stacktrace on java.lang.OutOfMemoryError?

**Reply by derekcollison** [reply]
Are you on latest server 2.10.7? 2.10.8 will be released tomorrow..

**Reply by yonigibbs** [reply]
The out of memory is of our own making (on purpose, chaos engineering as a test), in its own thread. So the stack trace there contains nothing of interest. In response to the error we cancel all the coroutines which are processing stream items. When they are cancelled they respond by NAKing the message they are currently processing. Additionally, we unsubscribe the NATS JetStream subscription.

We are currently just using the 2.10-alpine Docker image. I'll try tomorrow with the very latest Docker image and report back.

Thanks for the responses. Much appreciated!

**Reply by gregorycollins** [reply]
More details now that I have a moment to describe this better for you:

We create our streams as an archive stream (where messages are retained for a fixed period for auditing/replay) mirrored to a work queue stream:

```kotlin
    private fun addStream(
        name: String,
        subjects: String?,
        description: String?,
        retentionPolicy: RetentionPolicy?,
        block: (StreamConfiguration.Builder) -> StreamConfiguration.Builder,
    ) {
        val builder = StreamConfiguration.builder()
            .name(name)
            .replicas(config.getOrElse(natsCfg.replicas, 1))
            .storageType(StorageType.File)
            .discardPolicy(DiscardPolicy.Old)
        subjects?.let { builder.subjects(it) }
        description?.let { builder.description(it) }
        retentionPolicy?.let { builder.retentionPolicy(it) }
        val cfg = builder.let(block).build()
        logger.info("creating stream $name")
        if (preExistingStreams.contains(name)) {
            mgmt.updateStream(cfg)
        } else {
            mgmt.addStream(cfg)
        }
    }

    private fun addStreamWithArchive(
        name: String,
        subjects: String?,
        description: String?,
        retentionPolicy: RetentionPolicy?,
        maxAge: Duration = Duration.ofHours(24 * 30 * 3),
        build: (StreamConfiguration.Builder) -> StreamConfiguration.Builder,
    ) {
        addStream("$name-ARCHIVE", subjects, "archive for $description", RetentionPolicy.Limits) {
            it.maxAge(maxAge)
                .denyDelete(true)
        }
        addStream(name, null, description, retentionPolicy) {
            build(it)
                .mirror(
                    Mirror.builder()
                        .name("$name-ARCHIVE")
                        .build(),
                )
        }
    }
```

Every git version of our app creates its own durable Consumer of the work queue stream, listening to a `filteredSubject` sub-subject per version; e.g. if the stream subject is `foo.bar.>` then the workers subscribe to consumer `foo-bar-c-$v` listening to `foo.bar.$v`. We do this to implement a traffic shaping canary scheme that shunts k% of traffic to the canary on release (for various reasons we opted not to use nats subject mapping for this...). We log the sequence id of the messages both on the original `publish()` and in the worker app on pull (we are using `JetStreamSubscription.fetch()` on the pull side).

As Yoni mentioned, we use Kotlin coroutines -- these have issues with ThreadLocals sometimes, but I don't see any in nats.java besides a thread local RNG you use for wait jitter. As he also mentioned, we trap `CancellationException` on our worker coroutines, and if we see it, we NAK `Message`s in progress that haven't completed yet and cancel our subscription.

The exact behaviour we are seeing is:

1. the publisher sends O(100) messages to the versioned queue -- this is why Yoni's examples contain `UNKNOWN`, he is bench testing with version=UNKNOWN. we log the sequence number of the returned `PublishAck`
2. our app triggers self-cancellation, NAKs active traffic, and restarts. It eventually chews through all of the work items (except, sometimes, for one)
3. we've seen two scenarios: a) a stuck message which was never delivered to the worker at all, b) a stuck message which became stuck when we NAKed it during shutdown

Either way the symptoms are the same:
- `nats consumer info` reports a waiting pull
- new subscribers, either from our app or from the `natscli` command line, cannot pull the message
- the message is stuck forever (well, we've waited for 20x the `ackWait` time, which might as well be "forever" for our purposes), until the consumer is deleted and a new consumer is attached to the filter subject. The message we pull (on the new consumer we create after the old one is deleted) has a sequence number that is below the old ack floor.

We are really interested in concocting a minimal repro for you, in order to rule out PEBKAC. It'll be a few days.

Edit to say that we are running with nreplicas=1 in test, we will see if we can repro it with nreplicas=3

**Reply by gregorycollins** [reply]
P.S. our jnats client version is currently `2.16.9`, we could stand to fast-forward that if it would help

**Reply by derekcollison** [reply]
Looping in @scottf to comment on the java client lib.

**Reply by yonigibbs** [reply]
Upgraded the NATS Server Docker image to `nats:2.10.7-alpine` and the Jnats client to `2.17.2` (the latest version of both, I believe), and the issue persists. Also ran NATS server with `-DV` to get debug logging, but saw nothing obvious in there. Will now start building a new cut down project that hopefully we can replicate with, in order to rule out PEBKAC, as Greg says.

Quick question for the NATS folks: is there a way to look in the stream or consumer and figure out why a particular message isn't being picked up by the consumer? Any way we can get any metadata about the "stuck" message or anything like that? We've done `nats stream view` and `nats stream info` etc as included at the start of this conversation. Anything else we can try to help diagnose?

**Reply by scottf** [reply]
@yonigibbs Can you give me a sample of your fetch loop code? I'd like to see this in case anything in particular stands out and also so I can help you transition to our simplification api.
Also, can you give me a little more insight into the behavior you are seeing?

> our app triggers self-cancellation, NAKs active traffic, and restarts. It eventually chews through all of the work items (except, sometimes, for one)

By NAKs active traffic I'm assuming that you process any remaining items in the fetched list with NAKs.

What do you mean "(except, sometimes, for one)"?

**Reply by yonigibbs** [reply]
Thanks Scott. The code in the fetch loop is basically this (bearing in mind this is all Kotlin coroutine-y stuff):
```kotlin
// isAction is a property on `CoroutineScope` which is set to false when the coroutine is being cancelled
while (isActive) {
    val messages = natsSubscription.fetch(fetchBufferSize, Duration.ofMinutes(1))

    // We send these messages in a coroutine channel to be executed in parallel in a different coroutine
    for (message in messages) {
        channel.send(message)
    }
}
```

The code that picks up the messages from the coroutine channel is basically this:
```kotlin
val message = channel.receive()
try {
    // Do the work
    
    message.ack()
} catch (e: Throwable) {
    message.nak()
}
```

So there's multiple copies of those bits of code running concurrently (we have a thread pool consuming the queue). If any of them, or any other part of the system, hits an out-of-memory (which in our tests we're causing to happen on purpose) then we ask all the coroutines to end. This results in a `CancellationException` being thrown in each of them, being caught in each, and `message.nak()` being called on the message being processed in each coroutine.

Now, it is definitely possible that sometimes we will not NAK the message. For example say we poll and pick up some messages and just as we're about to put them in the coroutine channel the out-of-memory occurs, then we won't NAK. But in my view that's fine (in fact I think Greg was of the opinion that we shouldn't NAK at all after an OOM). Reason being that the app could hard crash at any point, so NATS can't expect us to NAK. If a consumer receives a message and then dies without NAK'ing it, NATS will eventually redeliver it to another consumer, and that's perfect.

And finally, when Greg said "except, sometimes, for one" what he's talking about is exactly the issue we're facing: say we put 100 messages in the queue. We then start our app. It processes, say, 20 of the items, then dies with the OOM. So we restart it, and it processes another 20 then dies, etc. After a few runs, it'll have processed every item in the queue. Except... we are pretty consistently seeing it leaving one (sometimes two) messages in the stream in this "stuck" state. By "stuck" I mean that the message is in the stream, but isn't being delivered to the consumer. And its sequence number seems to be lower than the `Ack floor` value.

**Reply by scottf** [reply]
Regarding NAK. As you probably know, when the server sends a message, it sets a timer (or a mark or something...) for the ack of that message (I think you have 2 minutes ack wait). 

* If the message is not ACKed in that time the server will resend to the consumer in whatever algorithm it does, not sure of exactly how it choses which message to send. It probably tries to resend as soon as possible.
* If the message is NACKed, this is similar to the message ack wait timing out. The client is saying "I can't process this so give it back to this consumer to try again". This is useful when you have multiple instances of the same consumer in a (consumer subscription) queue setup distributing the workload. Maybe the first instance is having a problem talking to the backend but the second instance will be fine.

The server tries to be smart about who it resends to since it is aware of this dynamic.

Its possible that the server resends right away, between your nack and your full consumer shutdown, depending on what pull requests are live. If the server sends it, it's out for 2 minutes (your ack wait) and might _seem_ like it's gone missing, unless you wait the full 2 minutes.

This is my current working hypothesis.

This might be affected by the old style fetch, which collects a bunch of messages then gives them to you, so the timing on the first message in the batch is affected by the collection of the messages. The timing on the later messages is affected by the processing time of all the messages before it.

Something that might help is moving to the simplification fetch. It works similar, except it does not block, meaning you can access the first message as soon as it's received instead of waiting for the full fetch batch to be received. (The old style iterate is like that as well.)

I would also then try to tune the fetch size and the ack wait to the size of your messages and to the time it generally takes to process each message.

**Reply by yonigibbs** [reply]
Thanks @scottf. It's definitely not the case that the "stuck" message just seems like it's gone missing. We've left the system for, like, an hour, then come back to it and connected the consumer and it's not picked up the message.

Re the "simplification fetch"... what is this exactly? Do you have a link to some docs or something?

**Reply by scottf** [reply]
Here are several examples: https://github.com/nats-io/nats.java/tree/main/src/examples/java/io/nats/examples/jetstream/simple
I would start with the ContextExample - the "Simplification" hangs off of the new ConsumerContext, which can be retrieved from a StreamContext or directly from a JetStreamContext
Once you are familiar with ConsumerContext (a wrapper around a real backend consumer and functions to consume messages using that consumer), you can look into the other examples in the simple directory.

Java Docs are here: https://javadoc.io/static/io.nats/jnats/2.17.2/io/nats/client/ConsumerContext.html
with most of the api in BaseConsumerContext https://javadoc.io/static/io.nats/jnats/2.17.2/io/nats/client/BaseConsumerContext.html

**Reply by gregorycollins** [reply]
I think we would prefer to move to an async fetch w/ `CompletableFuture` but we started building this app a year ago before you added some of those APIs. Unfortunately we can trigger the bug with a fetch block size of 1. We have actually finished the reproduction example for you, it's here: https://github.com/yonigibbs/nats-stuck-message

There are repro instructions which should hopefully be detailed enough for you to run the example. Running it for a few minutes triggers the condition reliably for both of us. Please take a look and let us know if you'd like us to file a bug ticket for it (or maybe you can do that if you can validate the reproduction?). I would be very relieved if the answer was "you are using this API wrong somehow and switching to do X instead will fix your issue". Thanks Derek/Scott/NATS-team 👍

**Reply by yonigibbs** [reply]
Did you get a chance to look at the repo that reproduces the issue? Any idea if this is just us doing something wrong, or a genuine bug somewhere in NATS? Want us to log an issue? Thanks!

**Reply by pdebakker-auguria** [reply]
Is there an update to this?
We are currently facing the same issue

**Reply by yonigibbs** [reply]
@pdebakker-auguria: myself and @gregorycollins had a long conversation on Slack with Scott, we supplied them with a couple of git repos that they could use to reproduce this issue, and last we heard (a couple of months back), the team had recognised this as a bug in the NATS server (I think) and it was on their backlog. I haven't heard since. Unfortunately for our project we had to back out of using NATS because of this issue, which is a shame, coz it has a lot of really great features we were using. Hopefully someone from the NATS team will be able to give you more info. Good luck!

**Reply by scottf** [reply]
It's a verified issue and it's being worked on. I don't know the current status, I've asked someone on the server team to reply.

**Reply by wallyqs** [reply]
This issue is something that we have improved in the v2.10.14 release (like in https://github.com/nats-io/nats-server/pull/5227).  Thanks again for putting together the @yonigibbs with the reproduction, I ran it with the latest version and this does not show up anymore using your test.
@pdebakker-auguria which version are you running on at the moment? please let us know if you still see this issue with the v2.10.14 version in your setup.

**Reply by pdebakker-auguria** [reply]
@wallyqs If I run into the issue again I will save the nats server version and let you know. Locally we just use the :latest version so I'm not quite sure on the version number.
Last week we also finally updated the version of our nats.go package.
Truth be told, I haven't seen the issue this week anymore, but I am hesitant to say it is gone completely because I've cried wolf about this before to my team (saying it was fixed because I hadn't seen the issue in a while and it then popping up again a day later).

But thank you guys for the hard work and the update here. It is much appreciated!

**Reply by wmxhw** [reply]
@wallyqs  I had a similar problem using the InterestPolicy，when send kill -9 to running nats.
nats log:
[WRN] Detected consumer '$G > {my stream name} > {my consumer name}' ack floor xxxx is ahead of stream's last sequence xxxx

nats-server version v2.10.14
golang client  github.com/nats-io/nats.go    v1.34.1


I tried modifying the configuration but it still didn’t work.

nats.conf:
...
jetstream {
    ...
    sync_interval: 5s
}
...


Thanks!

**Comment by niondir** [comment]
Maybe it's a similar issue here?

We are using nats Versions:
	github.com/nats-io/nats-server/v2 v2.10.22
	github.com/nats-io/nats.go v1.37.0


Consumer was not sending messages:
```
Last Delivered Message: Consumer sequence: 1,783,285 Stream sequence: 39,072,633
Acknowledgment Floor: Consumer sequence: 1,783,285 Stream sequence: 39,072,633
```
While the Stream has a sequence number of ~5m.

After recreating the consumer we got messages again and the stats were plausible again:
```
Last Delivered Message: Consumer sequence: 96 Stream sequence: 5,084,890 Last delivery: 3m42s ago
Acknowledgment Floor: Consumer sequence: 96 Stream sequence: 5,080,871 Last Ack: 3m39s ago
```

**Comment by xsteadfastx** [comment]
stepped into this problem after i just poweroff my testing vm today. one service just didnt consumed anymore. consumer was still there on the server and just removing it via natscli and restarting of the service brought it up again. so i wonder, would it be a fix to check on every service restart if there is a consumer with that name... and if.... delete it before creating?

**Reply by neilalexander** [reply]
Which server version?

**Reply by neilalexander** [reply]
It shouldn't be necessary to delete consumers to fix them, if they're breaking/getting stuck then we consider that a bug.

**Reply by xsteadfastx** [reply]
server version is 2.10.24 and nats.go version is 1.38.0. maybe through the poweroff something gets corrupted in the nats state.

**Reply by xsteadfastx** [reply]
i can even reproduce it. vm poweroff... i have to delete the consumer.

**Reply by neilalexander** [reply]
Can you please get `stream info` and `consumer info` when it is in a stuck state?

**Reply by xsteadfastx** [reply]
```
[root@localdev:/home/mpreuss]# nats stream info
? Select a Stream uplinkevent
Information for Stream uplinkevent created 2025-01-15 15:49:15

              Subjects: uplinkevent.*
              Replicas: 1
               Storage: File

Options:

             Retention: WorkQueue
       Acknowledgments: true
        Discard Policy: Old
      Duplicate Window: 2m0s
     Allows Msg Delete: true
          Allows Purge: true
        Allows Rollups: false

Limits:

      Maximum Messages: unlimited
   Maximum Per Subject: unlimited
         Maximum Bytes: unlimited
           Maximum Age: unlimited
  Maximum Message Size: unlimited
     Maximum Consumers: unlimited

State:

              Messages: 415
                 Bytes: 501 KiB
        First Sequence: 1 @ 2025-01-15 16:00:59
         Last Sequence: 415 @ 2025-01-15 16:04:49
      Active Consumers: 1
    Number of Subjects: 1
```

```
[root@localdev:/home/mpreuss]# nats consumer info
? Select a Stream uplinkevent
? Select a Consumer uplinkevent
Information for Consumer uplinkevent > uplinkevent created 2025-01-15T15:49:15+01:00

Configuration:

                    Name: uplinkevent
               Pull Mode: true
          Filter Subject: uplinkevent.*
          Deliver Policy: All
              Ack Policy: Explicit
                Ack Wait: 30.00s
           Replay Policy: Instant
         Max Ack Pending: 1,000
       Max Waiting Pulls: 512

State:

  Last Delivered Message: Consumer sequence: 903 Stream sequence: 1,976
    Acknowledgment Floor: Consumer sequence: 903 Stream sequence: 1,976
        Outstanding Acks: 0 out of maximum 1,000
    Redelivered Messages: 0
    Unprocessed Messages: 0
           Waiting Pulls: 1 of maximum 512
           ```

**Reply by neilalexander** [reply]
The consumer reports being ahead of the stream, which is why it is stalled. The `First sequence` and `Last sequence` of the stream, are they resetting when you do a kill scenario like this?

**Reply by xsteadfastx** [reply]
here is the nats stream info after i killed the consumer and recreated again. its working in this state.

```
[root@localdev:/home/mpreuss]# nats stream info
? Select a Stream uplinkevent
Information for Stream uplinkevent created 2025-01-15 15:49:15

              Subjects: uplinkevent.*
              Replicas: 1
               Storage: File

Options:

             Retention: WorkQueue
       Acknowledgments: true
        Discard Policy: Old
      Duplicate Window: 2m0s
     Allows Msg Delete: true
          Allows Purge: true
        Allows Rollups: false

Limits:

      Maximum Messages: unlimited
   Maximum Per Subject: unlimited
         Maximum Bytes: unlimited
           Maximum Age: unlimited
  Maximum Message Size: unlimited
     Maximum Consumers: unlimited

State:

              Messages: 0
                 Bytes: 0 B
        First Sequence: 1,343
         Last Sequence: 1,342 @ 2025-01-15 16:13:48
      Active Consumers: 1
```

```
[root@localdev:/home/mpreuss]# nats consumer info
? Select a Stream uplinkevent
? Select a Consumer uplinkevent
Information for Consumer uplinkevent > uplinkevent created 2025-01-15T16:06:25+01:00

Configuration:

                    Name: uplinkevent
               Pull Mode: true
          Filter Subject: uplinkevent.*
          Deliver Policy: All
              Ack Policy: Explicit
                Ack Wait: 30.00s
           Replay Policy: Instant
         Max Ack Pending: 1,000
       Max Waiting Pulls: 512

State:

  Last Delivered Message: Consumer sequence: 1,491 Stream sequence: 1,491 Last delivery: 777ms ago
    Acknowledgment Floor: Consumer sequence: 1,491 Stream sequence: 1,491 Last Ack: 777ms ago
        Outstanding Acks: 0 out of maximum 1,000
    Redelivered Messages: 0
    Unprocessed Messages: 0
           Waiting Pulls: 2 of maximum 512
           ```

**Reply by neilalexander** [reply]
Right, but what I'm asking is, are you observing the stream `First sequence` and `Last sequence` go backward in this scenario? i.e. `First sequence` resets back to 1.

**Reply by xsteadfastx** [reply]
oh yeah... its set to 

```
State:

              Messages: 24
                 Bytes: 28 KiB
        First Sequence: 1 @ 2025-01-15 16:18:55
         Last Sequence: 24 @ 2025-01-15 16:19:07
```

**Reply by xsteadfastx** [reply]
now i understand. removing and recreating the consumer also resets to sequence 1 so it starts to consume. im understanding but i have no idea why it should do it like that. so the stream loses it "state"?

**Reply by neilalexander** [reply]
It would appear to be. Is this a single-node setup or a cluster? Is `store_dir` set to a volume that doesn't get blown away? (i.e. it's not pointing to `/tmp` or something?)

**Reply by xsteadfastx** [reply]
single node setup. my store_dir is a `/var/lib/private/nats` set. systemd handled. i mean if i reboot the vm it keeps up with the sequences. so it only happens on poweroffs.

**Reply by neilalexander** [reply]
If you set `sync: always` in the `jetstream` block of the server config, does this improve things?

**Reply by xsteadfastx** [reply]
but the other stream is also strange. it consumes though...

```
[root@localdev:/home/mpreuss]# nats stream info
? Select a Stream recording
Information for Stream recording created 2025-01-15 16:22:33

              Subjects: recording.>
              Replicas: 1
               Storage: File

Options:

             Retention: Limits
       Acknowledgments: true
        Discard Policy: Old
      Duplicate Window: 2m0s
     Allows Msg Delete: true
          Allows Purge: true
        Allows Rollups: false

Limits:

      Maximum Messages: unlimited
   Maximum Per Subject: unlimited
         Maximum Bytes: unlimited
           Maximum Age: unlimited
  Maximum Message Size: unlimited
     Maximum Consumers: unlimited

State:

              Messages: 12,363
                 Bytes: 15 MiB
        First Sequence: 1 @ 2025-01-15 15:31:35
         Last Sequence: 12,363 @ 2025-01-15 16:40:01
      Active Consumers: 2
    Number of Subjects: 4
```

```
[root@localdev:/home/mpreuss]# nats consumer info
? Select a Stream recording
? Select a Consumer pusher-db
Information for Consumer recording > pusher-db created 2025-01-15T16:22:33+01:00

Configuration:

                    Name: pusher-db
               Pull Mode: true
         Filter Subjects: recording.recording.primary, recording.device_metadata, recording.device_errors
          Deliver Policy: All
              Ack Policy: Explicit
                Ack Wait: 30.00s
           Replay Policy: Instant
         Max Ack Pending: 1,000
       Max Waiting Pulls: 512

State:

  Last Delivered Message: Consumer sequence: 6,480 Stream sequence: 12,419 Last delivery: 24ms ago
    Acknowledgment Floor: Consumer sequence: 6,480 Stream sequence: 12,419 Last Ack: 24ms ago
        Outstanding Acks: 0 out of maximum 1,000
    Redelivered Messages: 0
    Unprocessed Messages: 0
           Waiting Pulls: 1 of maximum 512
```
why is the sequence still at 1? no i kinda getting nervous here ;-)

**Reply by xsteadfastx** [reply]
> If you set `sync: always` in the `jetstream` block of the server config, does this improve things?

testing it right now

**Reply by xsteadfastx** [reply]
first tests with sync always are positive. what does it do? i couldnt find anything in the docs.

**Comment by wallyqs** [comment]
Seems similar symptoms as in: https://github.com/nats-io/nats-server/issues/5412
